parts/linux/cloud-init/artifacts/cse_helpers.sh

```diff
@@ -384,40 +384,80 @@ should_skip_nvidia_drivers() {
     echo "$should_skip"
 }
 
-start_watch () {
-  capture_time=$(date +%s)
-  start_timestamp=$(date +%H:%M:%S)
+installJq () {
+  # jq is not available until downloaded in install-dependencies.sh with the installDeps function
+  # but it is needed earlier to call the capture_benchmarks function in pre-install-dependencies.sh
+  output=$(jq --version)
+  if [ -n "$output" ]; then
+    echo "$output"
+  else
+    if [[ $OS == $MARINER_OS_NAME ]]; then
+      sudo tdnf install -y jq
+      echo "jq was installed: $(jq --version)"
+    else
+      apt_get_install 5 1 60 jq
+      echo "jq was installed: $(jq --version)"
+    fi
+  fi
 }
 
-stop_watch () {
+capture_benchmarks () {
+  # the capture_benchmarks function is used to capture benchmark data for each section in VHD build scripts
+  # data will be stored in a JSON file and published as an artifact to be evaluated by github actions
+  set +x
+
+  local is_final_section=$1
+  local title=$2
 
   local current_time=$(date +%s)
   local end_timestamp=$(date +%H:%M:%S)
-  local difference_in_seconds=$((current_time - ${1}))
+  if [ "$is_final_section" = true ]; then
+    local difference_in_seconds=$((current_time - script_start_stopwatch))
+  else
+    local difference_in_seconds=$((current_time - section_start_stopwatch))
+  fi
 
   local elapsed_hours=$(($difference_in_seconds/3600))
   local elapsed_minutes=$((($difference_in_seconds%3600)/60))
   local elapsed_seconds=$(($difference_in_seconds%60))
+  printf -v total_time_elapsed "%02d:%02d:%02d" $elapsed_hours $elapsed_minutes $elapsed_seconds
+
+  # if the final section of the current script has not yet been reached, store the latest recorded benchmarking data in an array
+  if [[ ! "$is_final_section" == true ]]; then
+    benchmarks+=($title)
+    benchmarks+=($section_start_timestamp)
+    benchmarks+=($end_timestamp)
+    benchmarks+=($total_time_elapsed)
+  fi
   
-  printf -v benchmark "'${2}' - Total Time Elapsed: %02d:%02d:%02d" $elapsed_hours $elapsed_minutes $elapsed_seconds
-  if [ ${3} == true ]; then
-    printf -v start "     Start time: $script_start_timestamp"
-  else
-    printf -v start "     Start time: $start_timestamp"
+  if [[ "$is_final_section" == true ]]; then 
+    
+    # if current section is the final section of the current script, create a JSON object for the script as a whole
+    script_object=$(jq -n --arg script_name "$title" --arg script_start_timestamp "$script_start_timestamp" --arg end_timestamp "$end_timestamp" --arg total_time_elapsed "$total_time_elapsed" '{($script_name): {"overall": {"start_time": $script_start_timestamp, "end_time": $end_timestamp, "total_time_elapsed": $total_time_elapsed}}}')
+
+    #iterate over the benchmarks array in order to retrieve data from previous sections and create section objects for each
+    for ((i=0; i<${#benchmarks[@]}; i+=4)); do
```

Comment:
nit: consider using 4 arrays to store the information, namely, section_name, section_start_timestamp, end_timestamp, total_time_elapsed if there is no obvious benefit to use a single array. Just for easier debugging consideration. (Struct is not supported in bash script, otherwise it would be even better)